wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java

```diff
@@ -0,0 +1,102 @@
+import io.micrometer.core.instrument.binder.MeterBinder;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link MeterBinder} that is providing metrics about workspace successful or failed start time.
+ */
+@Singleton
+public class WorkspaceStartTrackerMeterBinder implements MeterBinder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(WorkspaceStartTrackerMeterBinder.class);
+
+  private final EventService eventService;
+
+  private final Map<String, Long> workspaceStartTime;
+  private Timer successTimer;
+  private Timer failTimer;
+
+  @Inject
+  public WorkspaceStartTrackerMeterBinder(EventService eventService) {
+    this(eventService, new ConcurrentHashMap<>());
+  }
+
+  @VisibleForTesting
+  WorkspaceStartTrackerMeterBinder(
+      EventService eventService, Map<String, Long> workspaceStartTime) {
+    this.eventService = eventService;
+    this.workspaceStartTime = workspaceStartTime;
+  }
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+
+    successTimer =
+        Timer.builder(workspaceMetric("start.time"))
+            .description("The time of workspace start")
+            .tags(withStandardTags("result", "success"))
+            .publishPercentileHistogram()
+            .sla(Duration.ofSeconds(10))
+            .minimumExpectedValue(Duration.ofSeconds(10))
+            .maximumExpectedValue(Duration.ofMinutes(15))
+            .register(registry);
+
+    failTimer =
+        Timer.builder(workspaceMetric("start.time"))
+            .description("The time of workspace start")
+            .tags(withStandardTags("result", "fail"))
+            .publishPercentileHistogram()
+            .sla(Duration.ofSeconds(10))
+            .minimumExpectedValue(Duration.ofSeconds(10))
+            .maximumExpectedValue(Duration.ofMinutes(15))
+            .register(registry);
+
+    // only subscribe to the event once we have the counters ready
+    eventService.subscribe(
+        event -> {
+          if (event.getPrevStatus() == WorkspaceStatus.STOPPED
+              && event.getStatus() == WorkspaceStatus.STARTING) {
+            workspaceStartTime.put(event.getWorkspaceId(), System.currentTimeMillis());
+          } else if (event.getPrevStatus() == WorkspaceStatus.STARTING) {
+            Long startTime = workspaceStartTime.remove(event.getWorkspaceId());
+            if (startTime != null) {
+              if (event.getStatus() == WorkspaceStatus.RUNNING) {
+                successTimer.record(Duration.ofMillis(System.currentTimeMillis() - startTime));
+              } else if (event.getStatus() == WorkspaceStatus.STOPPED) {
```

Comment:
Note that you will register failing even if a user interrupts workspace start.

Reply:
I consider this an acceptable scenario.